File: algorithms/algorithms-cplusplus/leetcode/Question_0117_Populating_Next_Right_Pointers_in_Each_Node_II.cc

// Definition for a Node.
class Node {
 public:
  int val;
  Node* left;
  Node* right;
  Node* next;

  Node() : val(0), left(NULL), right(NULL), next(NULL) {}

  Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

  Node(int _val, Node* _left, Node* _right, Node* _next)
    : val(_val), left(_left), right(_right), next(_next) {}
};

class Solution {
 public:
  Node* connect(Node* root) {
    Node dummy(0, NULL, NULL, NULL);
    Node *cur = root, *p = &dummy;
    while (cur) {
      if (cur->left) {
        p->next = cur->left;
        p = p->next;
      }
      if (cur->right) {
        p->next = cur->right;
        p = p->next;
      }
      cur = cur->next;
      if (!cur) {
        p = &dummy;
        cur = p->next;
        p->next = NULL;
      }
    }
    return root;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0117_Populating_Next_Right_Pointers_in_Each_Node_II.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0117_Populating_Next_Right_Pointers_in_Each_Node_II {
    // Definition for a Node.
    class Node {
        public int val;
        public Node left;
        public Node right;
        public Node next;
    
        public Node() {}
        
        public Node(int _val) {
            val = _val;
        }
    
        public Node(int _val, Node _left, Node _right, Node _next) {
            val = _val;
            left = _left;
            right = _right;
            next = _next;
        }
    };

    public Node connect(Node root) {
        Node cur = root, dummy = new Node(0, null, null, null), p = null;
        p = dummy;
        while (cur != null) {
            if (cur.left != null) {
                p.next = cur.left;
                p = p.next;
            }
            if (cur.right != null) {
                p.next = cur.right;
                p = p.next;
            }
            cur = cur.next;
            if (cur == null) {
                cur = dummy.next;
                dummy.next = null;
                p = dummy;
            }
        }
        return root;
    }
}
